mne_sandbox/preprocessing/dss.py

```
@@ -0,0 +1,108 @@
+         data_thresh=None, bias_thresh=None):
+    """Preprocess physiological data with denoising source separation (DSS)
+
+    Acts on covariance matrices; allows specification of arbitrary bias
+    functions (as compared to the public ``dss`` function, which forces the
+    bias function to be the evoked response).
+    """
+    data_eigval, data_eigvec = _pca(data_cov, data_n_keep, data_thresh)
+    W = np.diag(np.sqrt(1 / data_eigval))  # whitening matrix
+    # bias covariance projected into whitened PCA space of data channels
+    bias_white = W.T.dot(data_eigvec.T).dot(bias_cov).dot(data_eigvec).dot(W)
+    # proj. matrix from whitened data space to a space maximizing bias fxn
+    bias_eigval, bias_eigvec = _pca(bias_white, bias_n_keep, bias_thresh)
+    # proj. matrix from data to bias-maximizing space (DSS space)
+    dss_mat = data_eigvec.dot(W).dot(bias_eigvec)
+    # matrix to normalize DSS dimensions
+    N = np.diag(np.sqrt(1 / np.diag(dss_mat.T.dot(data_cov).dot(dss_mat))))
+    return dss_mat.dot(N)
+
+
+def _pca(cov, n_keep=None, thresh=None):
+    """Perform PCA decomposition
+
+    Parameters
+    ----------
+    cov : array-like
+        Covariance matrix
+    n_keep : int | None
+        Number of components to retain after decomposition.
+    thresh : float
+        Threshold (relative to the largest component) below which components
+        will be discarded.
+
+    Returns
+    -------
+    eigval : array
+        1-dimensional array of eigenvalues.
+    eigvec : array
+        2-dimensional array of eigenvectors.
+    """
+    eigval, eigvec = np.linalg.eig(cov)
+    eigval = np.abs(eigval)
+    sort_ix = np.argsort(eigval)[::-1]
+    eigvec = eigvec[:, sort_ix]
+    eigval = eigval[sort_ix]
+    if n_keep is not None:
+        eigval = eigval[:n_keep]
+        eigvec = eigvec[:, :n_keep]
+    if thresh is not None:
+        suprathresh = np.where(eigval / eigval.max() >= thresh)[0]
```

ahh this is a generic `_pca` function, so maybe throw the error in the calling function instead if necessary